### DISCUSSION TITLE
Is there a hook that allows adjusting eglot-workspace-configuration?

### DISCUSSION THREAD

**Discussion by antifuchs** [joaotavora/eglot#878] [Q&A] [answered]
I'm working with pyright across several projects, each of which has a `./venv` dir locally (mandated by pulumi - can't really set a "global" venv path that works for my other team members, too). Now, I could drop one `.dir-locals.el` where I set the `eglot-workspace-configuration` settings for venvPath and pythonPath into each project, but that is getting a bit tedious. I'd like to set those values programmatically for pyright connections.

Is there a hook for that, or a method I could specialize that would allow that kind of customization?

**Comment by joaotavora** [comment]
You _can_ set directory local variables programmatically. Look in the Emacs manual. If you can't find it, ping me here. Away from my dev box right now.

**Reply by antifuchs** [reply]
> You _can_ set directory local variables programmatically. Look in the Emacs manual. If you can't find it, ping me here. Away from my dev box right now.

Ah, yep, I'm familiar with [`dir-locals-set-class-variables` and `dir-locals-set-directory-class`](https://www.gnu.org/software/emacs/manual/html_node/emacs/Directory-Variables.html), but the problem is that the _values_ would need to be computed, and at that point I think I might have to set a `eval` dir-local setting (and eat the cost of having less-auditable dir local safety).

My thoughts were more around introducing something like `(cl-defgeneric eglot-workspace-configuration-values (server))`, like `eglot-initialization-options` - just used to compute the values sent to a server. I think custom server classes could wrap/override that method and provide additional settings for workspaces; that would be pretty neat for things like pyright's venv settings (which can be determined by looking at directories in the project just right).

What do you think? I'm happy to file a PR setting up that method if you agree.

**Comment by nemethf** [comment, marked as answer]
Can't you use relative paths in your global eglot-workspace-configuration? or in your pyrightconfig.json?

Also, the eglot-server-initialized-hook runs "before a connection was established".

**Reply by antifuchs** [reply]
Well, I'll be damned: A directory relative to the project root does work. I guess I was assuming this works the way `lsp-pyright` sets things up (which doesn't necessarily start the pyright server in the right directory & requires absolute pathnames due to that).

I still think it might be useful to have that method - instead of calling it from eglot-server-initialized-hook, I meant to extract the behavior around determining the value of workspace configuration settings from `(cl-defmethod eglot-handle-request (server (_method (eql workspace/configuration)) &key items))` into its own method and calling it from that one.

**Reply by joaotavora** [reply]
> I still think it might be useful to have that method 

I might be convinced :-).  Can you type two distinct solutions for your current problem?

1. how you do it now in your `.emacs` with `eglot-server-initialized-hook` as per @nemethf 's suggestion
2. how you would do it if you had access to the `eglot-workspace-configuration` generic function 

I'm curious to see which one is the more idiomatic.  I don't have anything "ideological" against 2 and I also do think it could be a good idea (and it could share the name with the variable: indeed the method for `t` could be the current behaviour which just returns the variable).  On the other hand, if 1. is reasonably idiomatic, then should we bother.
